README.md
# Bio Parsers
<!-- TOC -->

- [Bio Parsers](#bio-parsers)
  - [About this Repo](#about-this-repo)
  - [[CHANGELOG](CHANGELOG.md)](#changelogchangelogmd)
  - [Exported Functions](#exported-functions)
  - [Format Specification](#format-specification)
  - [Usage](#usage)
    - [install](#install)
    - [jsonToGenbank (same interface as jsonToFasta)](#jsontogenbank-same-interface-as-jsontofasta)
    - [anyToJson (same interface as genbankToJson, fastaToJson, xxxxToJson) (async required)](#anytojson-same-interface-as-genbanktojson-fastatojson-xxxxtojson-async-required)
    - [Options (for anyToJson or xxxxToJson)](#options-for-anytojson-or-xxxxtojson)
    - [ab1ToJson](#ab1tojson)
    - [snapgeneToJson (.dna files)](#snapgenetojson-dna-files)
    - [genbankToJson](#genbanktojson)
  - [Editing This Repo](#editing-this-repo)
    - [All collaborators:](#all-collaborators)
  - [Debug](#debug)
  - [Updating this repo](#updating-this-repo)
    - [Teselagen collaborators](#teselagen-collaborators)
    - [Outside collaborators](#outside-collaborators)
  - [Thanks/Collaborators](#thankscollaborators)

<!-- /TOC -->
## About this Repo
This repo contains a set of parsers to convert between datatypes through a generalized JSON format.

## [CHANGELOG](CHANGELOG.md) 

## Exported Functions
Use the following exports to convert to a generalized JSON format:
```
fastaToJson //handles fasta files (.fa, .fasta)
genbankToJson //handles genbank files (.gb, .gbk)
ab1ToJson //handles .ab1 sequencing read files 
sbolXmlToJson //handles .sbol files
snapgeneToJson //handles snapgene (.dna) files
anyToJson    //this handles any of the above file types based on file extension
```

Use the following exports to convert from a generalized JSON format back to a specific format:
```
jsonToGenbank
jsonToFasta
jsonToBed
```


## Format Specification
The generalized JSON format looks like:
```js
const generalizedJsonFormat = {
    "size": 25,
    "sequence": "asaasdgasdgasdgasdgasgdasgdasdgasdgasgdagasdgasdfasdfdfasdfa",
    "circular": true,
    "name": "pBbS8c-RFP",
    "description": "",
    "chromatogramData": { //only if parsing in an ab1 file
      "aTrace": [], //same as cTrace but for a
      "tTrace": [], //same as cTrace but for t
      "gTrace": [], //same as cTrace but for g
      "cTrace": [0,0,0,1,3,5,11,24,56,68,54,30,21,3,1,4,1,0,0, ...etc], //heights of the curve spaced 1 per x position (aka if the cTrace.length === 1000, then the max basePos can be is 1000)
      "basePos": [33, 46, 55, ...etc], //x position of the bases (can be unevenly spaced)
      "baseCalls": ["A", "T", ...etc],
      "qualNums": [],
    },
    "features": [
        {
            "name": "anonymous feature",
            "type": "misc_feature",
            "id": "5590c1978979df000a4f02c7", //Must be a unique id. If no id is provided, we'll autogenerate one for you
            "start": 1,
            "end": 3,
            "strand": 1,
            "notes": {},
        },
        {
            "name": "coding region 1",
            "type": "CDS",
            "id": "5590c1d88979df000a4f02f5",
            "start": 12,
            "end": 9,
            "strand": -1,
            "notes": {},
        }
    ],
}
```


## Usage
### install
`npm install -S bio-parsers` 

or 

`yarn add bio-parsers`

or 

use it from a script tag: 
```html
<script src="https://unpkg.com/bio-parsers/umd/bio-parsers.js"></script>
<script>
      async function main() {
        var jsonOutput = await window.bioParsers.genbankToJson(
          `LOCUS       kc2         108 bp    DNA     linear    01-NOV-2016
COMMENT             teselagen_unique_id: 581929a7bc6d3e00ac7394e8
FEATURES             Location/Qualifiers
     CDS             1..108
                     /label="GFPuv"
     misc_feature    61..108
                     /label="gly_ser_linker"
     bogus_dude      4..60
                     /label="ccmN_sig_pep"
     misc_feature    4..60
                     /label="ccmN_nterm_sig_pep"
                     /pragma="Teselagen_Part"
                     /preferred5PrimeOverhangs=""
                     /preferred3PrimeOverhangs=""
ORIGIN      
        1 atgaaggtct acggcaagga acagtttttg cggatgcgcc agagcatgtt ccccgatcgc
       61 ggtggcagtg gtagcgggag ctcgggtggc tcaggctctg ggg
//`
        );
        console.log('jsonOutput:', jsonOutput);
        var genbankString = window.bioParsers.jsonToGenbank(jsonOutput[0].parsedSequence);
        console.log(genbankString);
      }
      main();
</script>
```



see the `./umd_demo.html` file for a full working example

### jsonToGenbank (same interface as jsonToFasta)
```js
//To go from json to genbank:
import { jsonToGenbank } from "bio-parsers"
//You can pass an optional options object as the second argument. Here are the defaults
const options = {
  isProtein: false, //by default the sequence will be parsed and validated as type DNA (unless U's instead of T's are found). If isProtein=true the sequence will be parsed and validated as a PROTEIN type (seqData.isProtein === true)
  guessIfProtein: false, //if true the parser will attempt to guess if the sequence is of type DNA or type PROTEIN (this will override the isProtein flag)
  guessIfProteinOptions: {
    threshold = 0.90, //percent of characters that must be DNA letters to be considered of type DNA
    dnaLetters = ['G', 'A', 'T', 'C'] //customizable set of letters to use as DNA 
  }, 
  inclusive1BasedStart: false //by default feature starts are parsed out as 0-based and inclusive 
  inclusive1BasedEnd: false //by default feature ends are parsed out as 0-based and inclusive 
  // Example:
  // 0123456
  // ATGAGAG
  // --fff--  (the feature covers GAG)
  // 0-based inclusive start:
  // feature.start = 2
  // 1-based inclusive start:
  // feature.start = 3
  // 0-based inclusive end:
  // feature.end = 4
  // 1-based inclusive end:
  // feature.end = 5
} 
const genbankString = jsonToGenbank(generalizedJsonFormat, options)

```
### anyToJson (same interface as genbankToJson, fastaToJson, xxxxToJson) (async required)

```js
import { anyToJson } from "bio-parsers"

//note, anyToJson should be called using an await to allow for file parsing to occur (if a file is being passed)
const results = await anyToJson(
  stringOrFile, //if ab1 files are being passed in you should pass files only, otherwise strings or files are fine as inputs
  options //options.fileName (eg "pBad.ab1" or "pCherry.fasta") is important to pass here in order for the parser to!
) 

//we always return an array of results because some files my contain multiple sequences 
results[0].success //either true or false 
results[0].messages //either an array of strings giving any warnings or errors generated during the parsing process
results[0].parsedSequence //this will be the generalized json format as specified above :)
//chromatogram data will be here (ab1 only): 
results[0].parsedSequence.chromatogramData 
```

### Options (for anyToJson or xxxxToJson)
```js
//You can pass an optional options object as the third argument. Here are the defaults
const options = {
  fileName: "example.gb", //the filename is used if none is found in the genbank           
  isProtein: false, //if you know that it is a protein string being parsed you can pass true here
  parseFastaAsCircular: false; //by default fasta files are parsed as linear sequences. You can change this by setting parseFastaAsCircular=true 
  //genbankToJson options only
  inclusive1BasedStart: false //by default feature starts are parsed out as 0-based and inclusive 
  inclusive1BasedEnd: false //by default feature ends are parsed out as 0-based and inclusive 
  acceptParts: true //by default features with a feature.notes.pragma[0] === "Teselagen_Part" are added to the sequenceData.parts array. Setting this to false will keep them as features instead
}
```

### ab1ToJson

```js
import { ab1ToJson } from "bio-parsers"
const results = await ab1ToJson(
  //this can be either a browser file  <input type="file" id="input" multiple onchange="ab1ToJson(this.files[0])">
  // or a node file ab1ToJson(fs.readFileSync(path.join(__dirname, './testData/ab1/example1.ab1')));
  file, 
  options //options.fileName (eg "pBad.ab1" or "pCherry.fasta") is important to pass here in order for the parser to!
)

//we always return an array of results because some files my contain multiple sequences 
results[0].success //either true or false 
results[0].messages //either an array of strings giving any warnings or errors generated during the parsing process
results[0].parsedSequence //this will be the generalized json format as specified above :)
//chromatogram data will be here (ab1 only): 
results[0].parsedSequence.chromatogramData 
```

### snapgeneToJson (.dna files) 
```js
import { snapgeneToJson } from "bio-parsers"
//file can be either a browser file  <input type="file" id="input" multiple onchange="snapgeneToJson(this.files[0])">
// or a node file snapgeneToJson(fs.readFileSync(path.join(__dirname, './testData/ab1/example1.ab1')));
const results = await snapgeneToJson(file,options)
```

### genbankToJson

```js
import { genbankToJson } from "bio-parsers"

const result = genbankToJson(string, options)

console.info(result)
// [
//     {
//         "messages": [
//             "Import Error: Illegal character(s) detected and removed from sequence. Allowed characters are: atgcyrswkmbvdhn",
//             "Invalid feature end:  1384 detected for Homo sapiens and set to 1",
//         ],
//         "success": true,
//         "parsedSequence": {
//             "features": [
//                 {
//                     "notes": {
//                         "organism": [
//                             "Homo sapiens"
//                         ],
//                         "db_xref": [
//                             "taxon:9606"
//                         ],
//                         "chromosome": [
//                             "17"
//                         ],
//                         "map": [
//                             "17q21"
//                         ]
//                     },
//                     "type": "source",
//                     "strand": 1,
//                     "name": "Homo sapiens",
//                     "start": 0,
//                     "end": 1
//                 }
//             ],
//             "name": "NP_003623",
//             "sequence": "gagaggggggttatccccccttcgtcagtcgatcgtaacgtatcagcagcgcgcgagattttctggcgcagtcag",
//             "circular": true,
//             "extraLines": [
//                 "DEFINITION  contactin-associated protein 1 precursor [Homo sapiens].",
//                 "ACCESSION   NP_003623",
//                 "VERSION     NP_003623.1  GI:4505463",
//                 "DBSOURCE    REFSEQ: accession NM_003632.2",
//                 "KEYWORDS    RefSeq."
//             ],
//             "type": "DNA",
//             "size": 925
//         }
//     }
// ]

```

You can see more examples by looking at the tests.



## Editing This Repo
### All collaborators:
Edit/create a new file and update/add any relevant tests.
Make sure they pass by running `yarn test`

## Debug
```
yarn test-debug
```

## Updating this repo
### Teselagen collaborators
Commit and push all changes
Sign into npm using the teselagen npm account (npm whoami)

```
npm version patch|minor|major
npm publish
```


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v8.3.12](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.3.11...v8.3.12)

- adding a umd example to the readme and updating the validation function slightly [`849b802`](https://github.com/TeselaGen/ve-sequence-parsers/commit/849b802f4307926d6788b79d290a9a32a0aa55ef)

#### [v8.3.11](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.3.10...v8.3.11)

> 1 June 2021

- removing errant console [`5de90ab`](https://github.com/TeselaGen/ve-sequence-parsers/commit/5de90ab7d33628a7319f649c719b6d4a91b26f69)

#### [v8.3.10](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.3.6...v8.3.10)

> 28 May 2021

- keep casing, add back in oligo option [`791bb96`](https://github.com/TeselaGen/ve-sequence-parsers/commit/791bb96f9f64aac1c96101d8414044fc3df6559f)
- fix genbank DEFINITION parsing to handle new lines + test [`ac7ffa2`](https://github.com/TeselaGen/ve-sequence-parsers/commit/ac7ffa231eeac4888575657c5332ae0e361512a4)
- Do not create uracil features. Always change u's to t's [`6cc7c0f`](https://github.com/TeselaGen/ve-sequence-parsers/commit/6cc7c0fd2b9016240478850b285f782e3517f79a)

#### [v8.3.6](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.3.5...v8.3.6)

> 10 May 2021

- updating jsonToFasta to better handle isProtein option [`a733821`](https://github.com/TeselaGen/ve-sequence-parsers/commit/a7338217b29890243493f51514fb16cce53f8291)
- Merge pull request #178 from TeselaGen/dependabot/npm_and_yarn/hosted-git-info-2.8.9 [`b41cfec`](https://github.com/TeselaGen/ve-sequence-parsers/commit/b41cfec1ded1bdf1e7815c6f1a1b17424f26fe10)
- [Security] Bump hosted-git-info from 2.8.4 to 2.8.9 [`ab7424e`](https://github.com/TeselaGen/ve-sequence-parsers/commit/ab7424e30992a7019b60016bcfee4a0566e387eb)

#### [v8.3.5](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.3.4...v8.3.5)

> 13 April 2021

- fixing logical error in longestFeatureTypeLength calculation [`6b7a624`](https://github.com/TeselaGen/ve-sequence-parsers/commit/6b7a6242d4f97a3a96df176033dafad6aa5a6666)

#### [v8.3.4](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.3.3...v8.3.4)

> 13 April 2021

- fixing jsonToGenbank to coerce primers with type 'primer' to type 'primer_bind' [`06e5424`](https://github.com/TeselaGen/ve-sequence-parsers/commit/06e5424043d9061fcb13360bc672a6d08c81e59a)

#### [v8.3.3](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.3.0...v8.3.3)

> 13 April 2021

- Rename primer features to primer_bind on export [`ebc286e`](https://github.com/TeselaGen/ve-sequence-parsers/commit/ebc286e923a47c2ab91a90ad699e023548777d84)
- Fix isBrowser logic [`51e1e9d`](https://github.com/TeselaGen/ve-sequence-parsers/commit/51e1e9d5771c0c5590141a110a1bfd18aa4e79a9)

#### [v8.3.0](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.2.0...v8.3.0)

> 2 April 2021

- making getUtf8StringFromFile work in a node env; removing ability to be able to parse gff or sbol without specifying a fileName (for the file extension) [`4771a20`](https://github.com/TeselaGen/ve-sequence-parsers/commit/4771a200d2ca563e99c74a8a93c8e0445b10b52c)

#### [v8.2.0](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.1.0...v8.2.0)

> 1 April 2021

- improving robustness of nodejs arraybuffer parsing [`1d461d7`](https://github.com/TeselaGen/ve-sequence-parsers/commit/1d461d76104661e7805db099ea4ae15310bd5a64)

#### [v8.1.0](https://github.com/TeselaGen/ve-sequence-parsers/compare/v8.0.0...v8.1.0)

> 31 March 2021

- updating snapgeneToJson to follow same format as other parsers and not throw an error but instead just return a non-successful result [`6039a3d`](https://github.com/TeselaGen/ve-sequence-parsers/commit/6039a3d25973bae729a41f827fec8189b49c9a16)

### [v8.0.0](https://github.com/TeselaGen/ve-sequence-parsers/compare/v7.0.13...v8.0.0)

> 31 March 2021

- removing callback style functions; only async/await and promises supported now [`c04762f`](https://github.com/TeselaGen/ve-sequence-parsers/commit/c04762f122e909a43d5283451a335d97fce35368)

#### [v7.0.13](https://github.com/TeselaGen/ve-sequence-parsers/compare/v7.0.12...v7.0.13)

> 30 March 2021

- fixing regression that removed the ability to pass a string to anyToJson while in a browser context [`a590278`](https://github.com/TeselaGen/ve-sequence-parsers/commit/a59027822cbfd44ac3ae5fd81df94a7a59b45a56)
- Merge pull request #172 from TeselaGen/dependabot/npm_and_yarn/y18n-3.2.2 [`f2491ed`](https://github.com/TeselaGen/ve-sequence-parsers/commit/f2491edb6adbc2249111d5d49f8ffcc78bc7cad7)
- [Security] Bump y18n from 3.2.1 to 3.2.2 [`d73a5a5`](https://github.com/TeselaGen/ve-sequence-parsers/commit/d73a5a5fe0db7491b88284fc5475302079c704f2)

#### [v7.0.12](https://github.com/TeselaGen/ve-sequence-parsers/compare/v7.0.11...v7.0.12)

> 12 March 2021

- removing /test files from npm [`9bb1eaf`](https://github.com/TeselaGen/ve-sequence-parsers/commit/9bb1eaf32b5fbbff2139b23abfc0e4af20973289)

#### [v7.0.11](https://github.com/TeselaGen/ve-sequence-parsers/compare/v7.0.10...v7.0.11)

> 12 March 2021

- removing /test files from npm [`696823e`](https://github.com/TeselaGen/ve-sequence-parsers/commit/696823ea6130a6339a8c5032b04b4a286413d40e)

#### [v7.0.10](https://github.com/TeselaGen/ve-sequence-parsers/compare/v7.0.9...v7.0.10)

> 12 March 2021

- removing /test files from npm [`ea95cf5`](https://github.com/TeselaGen/ve-sequence-parsers/commit/ea95cf55edc6b8e88f113ad4128e6efa9f6cb595)


genbankToJson.js
var flattenSequenceArray = require('./utils/flattenSequenceArray');
var validateSequenceArray = require('./utils/validateSequenceArray');
// var some = require('lodash/collection/some');
var splitStringIntoLines = require('./utils/splitStringIntoLines.js');

function parseGenbankFileToOurOldTeselagenDataType(string, onFileParsedUnwrapped, options) {
    var onFileParsed = function(sequences, options) { //before we call the onFileParsed callback, we need to flatten the sequence, and convert the old sequence data to the new data type
        onFileParsedUnwrapped(validateSequenceArray(flattenSequenceArray(sequences), options));
    };
    options = options || {}
    var inclusive1BasedStart = options.inclusive1BasedStart
    var inclusive1BasedEnd = options.inclusive1BasedEnd
    
    var resultsArray = [];
    var result;
    var currentFeatureNote;

    var genbankAnnotationKey = {
        LOCUS_TAG: "LOCUS",
        DEFINITION_TAG: "DEFINITION",
        ACCESSION_TAG: "ACCESSION",
        VERSION_TAG: "VERSION",
        KEYWORDS_TAG: "KEYWORDS",
        //SEGMENT_TAG:"SEGMENT"
        SOURCE_TAG: "SOURCE",
        ORGANISM_TAG: "ORGANISM",
        REFERENCE_TAG: "REFERENCE",
        AUTHORS_TAG: "AUTHORS",
        CONSORTIUM_TAG: "CONSRTM",
        TITLE_TAG: "TITLE",
        JOURNAL_TAG: "JOURNAL",
        PUBMED_TAG: "PUBMED",
        REMARK_TAG: "REMARK",
        COMMENT_TAG: "COMMENT",
        FEATURES_TAG: "FEATURES",
        BASE_COUNT_TAG: "BASE COUNT",
        //CONTIG_TAG: "CONTIG"
        ORIGIN_TAG: "ORIGIN",
        END_SEQUENCE_TAG: "//"
    };

    try {
        var lines = splitStringIntoLines(string);
        var LASTTYPE = false;
        var featureLocationIndentation;

        if (lines === null) {
            addMessage("Import Error: Sequence file is empty");
        }

        lines.some(function(line) {
            if (line === null) {
                return true; //break the some loop
            }
            var key = getLineKey(line);
            var val = getLineVal(line);
            var isKeyRunon = isKeywordRunon(line);
            var isSubKey = isSubKeyword(line);
            var isKey = isKeyword(line);

            //only set a new lasttype in the case that we've encountered a key that warrants it. 
            if (key === "LOCUS") {
                LASTTYPE = key;
            }
            else if (key === "REFERENCE") {
                LASTTYPE = key;
            }
            else if (key === "FEATURES") {
                LASTTYPE = key;
            }
            else if (key === "ORIGIN") {
                LASTTYPE = key;
            }
            else if (key === "//") {
                LASTTYPE = key;
            }
            else if (isKey === true) {
                LASTTYPE = key;
            }

            // IGNORE LINES: DO NOT EVEN PROCESS
            if (line.trim() === "" || key === ";") {
                //tnr: don't add the following message because it is not particularly informative
                // addMessage(
                //     "Warning: Empty line, or ';' detected. Ignoring line: " +
                //     line);
                return false; // go to next line
            }


            switch (LASTTYPE) {
                case genbankAnnotationKey.LOCUS_TAG:
                    parseLocus(line, key, val);
                    break;
                case genbankAnnotationKey.FEATURES_TAG:
                    //If no location is specified, exclude feature and return messages
                    if (val === "") {
                        addMessage("Warning: The feature '" + key +
                            "'' has no location specified. This line has been ignored: line" +
                            line);
                        break;
                    }
                    parseFeatures(line, key, val);
                    break;
                case genbankAnnotationKey.ORIGIN_TAG:
                    parseOrigin(line, key);
                    break;
                case genbankAnnotationKey.END_SEQUENCE_TAG:
                    endSeq();
                    break;
                case "COMMENT":
                    // do nothing
                    // console.warn("GenbankManager.lineParser(: This line contains a 'COMMENT' and has been ignored: " + line);
                    extractExtraLine(line);
                    break;
                default: // FOLLOWING FOR KEYWORDS NOT PREVIOUSLY DEFINED IN CASES
                    extractExtraLine(line);
                    if (key === "BASE") {
                        // do nothing;              // BLANK LINES || line with ;;;;;;;;;  || "BASE COUNT"
                        // console.warn("Parsing GenBank File: This line with BaseCount has been ignored: " + line);
                        addMessage(
                            "Warning: This BaseCount line has been ignored: " + line);
                        break;
                    }
                    else if (isKey) {
                        // REGULAR KEYWORDS (NOT LOCUS/FEATURES/ORIGIN) eg VERSION, ACCESSION, SOURCE, REFERENCE
                        // lastObj = parseKeyword(line, gb);
                    }
                    else if (isSubKey) { // REGULAR SUBKEYWORD, NOT FEATURE eg AUTHOR, ORGANISM
                        // tmp = gb.getLastKeyword();
                        // lastObj = parseSubKeyword(tmp, line, gb);
                    }
                    else if (isKeyRunon) { // RUNON LINES FOR NON-FEATURES
                        // lastObj.setValue(lastObj.getValue() + Teselagen.StringUtil.rpad("\n"," ",13) + Ext.String.trim(line));
                        // lastObj.appendValue(Teselagen.StringUtil.rpad("\n"," ",13) + Ext.String.trim(line), gb);
                    }
                    else {
                        // console.warn("Parsing GenBank File: This line has been ignored: " + line);
                        addMessage("Warning: This line has been ignored: " + line);
                    }
            }
        });
    }
    catch (e) {
        //catch any errors and set the result
        result = {
            success: false,
            messages: ['Import Error: Invalid File']
        };
    }

    //catch the case where we've successfully started a sequence and parsed it, but endSeq isn't called correctly
    if (result.success &&
        resultsArray[resultsArray.length - 1] !== result) {
        //current result isn't in resultsArray yet
        //so we call endSeq here
        endSeq();
    }
    //call the callback
    onFileParsed(resultsArray, options);

    function newSeq() {
        result = {
            messages: [],
            success: true,
            parsedSequence: {
                features: [],
                name: '',
                // inData: {},
                sequence: ''
            }
        };
    }

    function endSeq() {
        //do some post processing clean-up
        postProcessCurSeq();
        //push the result into the resultsArray
        resultsArray.push(result);
    }

    function getCurrentFeature() {
        return result.parsedSequence.features[result.parsedSequence.features.length - 1];
    }

    function addMessage(msg) {
        return result.messages.push(msg);
    }

    function postProcessCurSeq() {
        if (result.parsedSequence && result.parsedSequence.features) {
            for (var i = 0; i < result.parsedSequence.features.length; i++) {
                result.parsedSequence.features[i] = postProcessGenbankFeature(result.parsedSequence.features[i]);
            }
        }
    }

    function parseOrigin(line, key) {
        if (key !== genbankAnnotationKey.ORIGIN_TAG) {
            var new_line = line.replace(/[\s]*[0-9]*/g, "");
            result.parsedSequence.sequence += new_line;
        }
    }

    function parseLocus(line) {
        newSeq();
        var locusName
        var linear
        var date
        var lineArr = line.split(/[\s]+/g);

        if (lineArr.length <= 1) {
            console.warn("Parsing GenBank File: WARNING! Locus line contains no values!");
            // TODO
            addMessage("Import Warning: Locus line contains no values: " + line);
        }
        locusName = lineArr[1];

        // Linear vs Circular?
        linear = true;
        for (var i = 1; i < lineArr.length; i++) {
            if (lineArr[i].match(/circular/gi)) {
                linear = false;
            }
        }

        // Date and Div
        // Date is in format:1-APR-2012
        for (var j = 1; j < lineArr.length; j++) {
            if (lineArr[j].match(/-[A-Z]{3}-/g)) {
                date = lineArr[j];
            }
            //tnr: not sure what this is supposed to be doing..
            // if (lineArr[j].match(/^[A-Z]{3}/g) && lineArr[j].length === 3 && !lineArr[j].match(/DNA|RNA/g)) {
            //     div = lineArr[j];
            // }
        }
        result.parsedSequence.name = locusName;
        result.parsedSequence.date = date;
        result.parsedSequence.circular = !linear;
    }

    function extractExtraLine(line) {
        if (result.parsedSequence) {
            if (!result.parsedSequence.extraLines) {
                result.parsedSequence.extraLines = [];
            }
            result.parsedSequence.extraLines.push(line);
        }
        else {
            throw ('no sequence yet created upon which to extract an extra line!');
        }
    }
    var lastLineWasFeaturesTag;
    var lastLineWasLocation;
    function parseFeatures(line, key, val) {
        var strand;
        // FOR THE MAIN FEATURES LOCATION/QUALIFIER LINE
        if (key === genbankAnnotationKey.FEATURES_TAG) {
            lastLineWasFeaturesTag = true;
            return;
        }

        if (lastLineWasFeaturesTag) {
            //we need to get the indentation of feature locations
            featureLocationIndentation = getLengthOfWhiteSpaceBeforeStartOfLetters(line);
            //set lastLineWasFeaturesTag to false 
            lastLineWasFeaturesTag = false;
        }

        // FOR LOCATION && QUALIFIER LINES
        if (isFeatureLineRunon(line, featureLocationIndentation)) {
            //the line is a continuation of the above line
            if (lastLineWasLocation) {
                //the last line was a location, so the run-on line is expected to be a feature location as well
                parseFeatureLocation(line.trim());
                lastLineWasLocation = true;
            }
            else {
                //the last line was a note
                if (currentFeatureNote) {
                    //append to the currentFeatureNote
                    currentFeatureNote[currentFeatureNote.length - 1] += line.trim().replace(/\"/g, "");
                }
                lastLineWasLocation = false;
            }
        }
        else { // New Element/Qualifier lines. Not runon lines.
            if (isNote(line)) { // is a new Feature Element (e.g. source, CDS) in the form of  "[\s] KEY  SEQLOCATION"
                // is a FeatureQualifier in the /KEY="BLAH" format; could be multiple per Element
                //Check that feature did not get skipped for missing location
                if (getCurrentFeature()) {
                    parseFeatureNote(line);
                    lastLineWasLocation = false;
                }
                else {
                    return;
                }
            }
            else {
                //the line is a location, so we make a new feature from it
                if (val.match(/complement/g)) {
                    strand = -1;
                }
                else {
                    strand = 1;
                }

                newFeature();
                var feat = getCurrentFeature();
                feat.type = key;
                feat.strand = strand;

                parseFeatureLocation(val);
                lastLineWasLocation = true;
            }
        }

    }

    function newFeature() {
        result.parsedSequence.features.push({
            locations: [],
            notes: {}
        });
    }

    function isNote(line) {
        var qual = false;
        /*if (line.charAt(21) === "/") {//T.H. Hard coded method
           qual = true;
         }*/
        if (line.trim().charAt(0).match(/\//)) { // searches based on looking for / in beginning of line
            qual = true;
        }
        else if (line.match(/^[\s]*\/[\w]+=[\S]+/)) { // searches based on "   /key=BLAH" regex
            qual = true;
        }
        return qual;
    }


    function parseFeatureLocation(locStr) {
        locStr = locStr.trim();
        var locArr = [];
        locStr.replace(/(\d+)/g, function(string, match) {
            locArr.push(match);
        });
        for (var i = 0; i < locArr.length; i += 2) {
            var start = parseInt(locArr[i]) - (inclusive1BasedStart ? 0 : 1);
            var end = parseInt(locArr[i + 1]) - (inclusive1BasedEnd ? 0 : 1);
            if (isNaN(end)) { 
                //if no end is supplied, assume that the end should be set to whatever the start is
                //this makes a feature location passed as:
                //147
                //function like:
                //147..147
                end = start
            }
            var location = {
                start: start,
                end: end
            };
            var feat = getCurrentFeature();
            feat.locations.push(location);
        }
    }

    function parseFeatureNote(line) {
        var newLine, lineArr;

        newLine = line.trim();
        newLine = newLine.replace(/^\/|"$/g, "");
        lineArr = newLine.split(/=\"|=/);

        var val = lineArr[1];

        if (val) {
            val = val.replace(/\\/g, " ");

            if (line.match(/=\"/g)) {
                val = val.replace(/\".*/g, "");
            }
            else if (val.match(/^\d+$/g)) {
                val = parseInt(val);
            }
        }
        var key = lineArr[0];
        var currentNotes = getCurrentFeature().notes;
        if (currentNotes[key]) {
            //array already exists, so push value into it
            currentNotes[key].push(val);
        }
        else {
            //array doesn't exist yet, so create it and populate it with the value
            currentNotes[key] = [val];
        }
        currentFeatureNote = currentNotes[key];
    }

    function getLineKey(line) {
        var arr;
        line = line.replace(/^[\s]*/, "");

        if (line.indexOf("=") < 0) {
            arr = line.split(/[\s]+/);
        }
        else {
            arr = line.split(/=/);
        }

        return arr[0];
    }

    function getLineVal(line) {
        var arr;

        if (line.indexOf("=") < 0) {
            line = line.replace(/^[\s]*[\S]+[\s]+|[\s]+$/, "");
            line = line.trim();
            return line;
        }
        else {
            arr = line.split(/=/);
            return arr[1];
        }
    }

    function isKeyword(line) {
        var isKey = false;
        if (line.substr(0, 10).match(/^[\S]+/)) {
            isKey = true;
        }
        return isKey;
    }

    function isSubKeyword(line) {
        var isSubKey = false;
        if (line.substr(0, 10).match(/^[\s]+[\S]+/)) {
            isSubKey = true;
        }
        return isSubKey;
    }

    function isKeywordRunon(line) {
        var runon;
        if (line.substr(0, 10).match(/[\s]{10}/)) {
            runon = true;
        }
        else {
            runon = false;
        }
        return runon;
    }

    function postProcessGenbankFeature(feat) {
        if (feat.notes.label) {
            feat.name = feat.notes.label[0];
        }
        else if (feat.notes.gene) {
            feat.name = feat.notes.gene[0];
        }
        else if (feat.notes.ApEinfo_label) {
            feat.name = feat.notes.ApEinfo_label[0];
        }
        else if (feat.notes.name) {
            feat.name = feat.notes.name[0];
        }
        else if (feat.notes.organism) {
            feat.name = feat.notes.organism[0];
        }
        else if (feat.notes.locus_tag) {
            feat.name = feat.notes.locus_tag[0];
        }
        else if (feat.notes.note) {
            //if the name is coming from a note, shorten the name to 100 chars long
            feat.name = feat.notes.note[0].substr(0,100);
        }
        else {
            feat.name = '';
        }
        //shorten the name to a reasonable length if necessary and warn the user about it
        var oldName = feat.name;
        feat.name = feat.name.substr(0,100);
        if (feat.name !== oldName) {
            addMessage('Warning: Shortening name of sequence (max 100 chars)')
        }
        return feat;
    }
}

function isFeatureLineRunon(line, featureLocationIndentation) {
    var indentationOfLine = getLengthOfWhiteSpaceBeforeStartOfLetters(line);
    if (featureLocationIndentation === indentationOfLine) {
        //the feature location indentation calculated right after the feature tag 
        //cannot be the same as the indentation of the line
        //
        //FEATURES             Location/Qualifiers
        //     rep_origin      complement(1074..3302)
        //01234  <-- this is the indentation we're talking about     
        return false; //the line is NOT a run on
    }

    var trimmed = line.trim();
    if (trimmed.charAt(0).match(/\//)) {
        //the first char in the trimmed line cannot be a /
        return false; //the line is NOT a run on
    }
    //the line is a run on
    return true;
    //run-on line example:
    //FEATURES             Location/Qualifiers
    //     rep_origin      complement(1074..3302)
    //                 /label=pSC101**
    //                 /note="REP_ORIGIN REP_ORIGIN pSC101* aka pMPP6, gives plasm
    //                 id number 3 -4 copies per cell, BglII site in pSC101* ori h <--run-on line!
    //                 as been dele ted by quick change agatcT changed to agatcA g <--run-on line!
    //                 iving pSC101* * pSC101* aka pMPP6, gives plasmid number 3-4 <--run-on line!
    //                 copies p er cell, BglII site in pSC101* ori has been delet  <--run-on line!
    //                 ed by quic k change agatcT changed to agatcA giving pSC101* <--run-on line!
    //                 * [pBbS0a-RFP]"                                             <--run-on line!
    //                 /gene="SC101** Ori"
    //                 /note="pSC101* aka pMPP6, gives plasmid number 3-4 copies p
    //                 er cell, BglII site in pSC101* ori has been deleted by qui
    //                 c k change agatcT changed to agatcA giving pSC101**"
    //                 /vntifkey="33"
}

function getLengthOfWhiteSpaceBeforeStartOfLetters(string) {
    var match = /^\s*/.exec(string);
    if (match !== null) {
        return (match[0].length);
    }
    else {
        return 0;
    }
}

module.exports = parseGenbankFileToOurOldTeselagenDataType;
